### PR TITLE
feat: binary field

### PR DESCRIPTION
**Pull request by lonerapier**
It changes the following:

- binary field
- binary field extensions using tower field arithmetic

I've kept the implementation closer to theory as described in Section 2.3 of [Binius](https://eprint.iacr.org/2023/1784.pdf). let me know if anything needs more work. 

TODO:
- [ ] figure out `primitive_element_generator` in extension fields
- [x] docs and tests
- [x] demonstrate power of binary fields by showing examples of embedding and small-by-large multiplication as described in the paper

### CODE REVIEW

File: src/field/binary/extension.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+use super::BinaryField;
+use crate::field::FiniteField;
+
+/// Binary extension field GF_{2^{2^K}} using binary towers arithmetic as explained in Section 2.3 of [DP23b](https://eprint.iacr.org/2023/1784.pdf)
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BinaryFieldExtension<const K: usize>
+where [(); 1 << K]: {
+  /// coefficients of field element
+  pub coefficients: [BinaryField; 1 << K],
+}
```
Comment (Autoparallel) [inline src/field/binary/extension.rs:16]: This guy is a const expr fiend now!!

File: src/field/binary/mod.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+};
+
+use crate::field::FiniteField;
+
+pub mod extension;
+#[cfg(test)] pub mod tests;
+
+/// binary field containing element `{0,1}`
+#[derive(Debug, Default, Hash, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BinaryField(u8);
```
Comment (Autoparallel) [inline src/field/binary/mod.rs:14]: Perhaps a type alias instead with:
```suggestion
pub type BinaryField = PrimeField<2>;
```

Comment (Autoparallel) [reply src/field/binary/mod.rs:14]: I believe that would collapse a lot of this module here.

File: src/field/binary/extension.rs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+  iter::{Product, Sum},
+  ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Rem, Sub, SubAssign},
+};
+
+use super::BinaryField;
+use crate::field::FiniteField;
+
+/// Binary extension field GF_{2^{2^K}} using binary towers arithmetic as explained in Section 2.3 of [DP23b](https://eprint.iacr.org/2023/1784.pdf)
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BinaryFieldExtension<const K: usize>
```
Comment (Autoparallel) [inline src/field/binary/extension.rs:12]: I think if you take the type I suggested, you could basically work with this and think of it like a nested set of very specifically chosen `GaloisField<2,2>`. Almost like a `GaloiseField<GaloisField<...>>` type of object (I know this latter type would not compile, but hopefully it gets the point across).

File: src/field/mod.rs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 use super::*;
 
+pub mod binary;
```
Comment (Autoparallel) [inline src/field/mod.rs:5]: perhaps tower here instead?

**Review by Autoparallel — COMMENTED** [review]
This is looking good my friend!!!

Suggestions so far only. Hopefully nothing nitpicky :)

File: src/field/binary/mod.rs (hunk 2 above)
Comment (lonerapier) [reply src/field/binary/mod.rs:14]: that'll work flawlessly, but i believe it wouldn't allow us to showcase the main properties of binary field, i.e. addition being bitwise XOR and multiplication bitwise AND.

File: src/field/binary/extension.rs (hunk 1 above)
Comment (lonerapier) [reply src/field/binary/extension.rs:16]: everywhere I go, I see compile time constants now :)

**Comment by lonerapier** [comment]
@0xJepsen @Autoparallel this is ready for review now, I have added binary fields using our ronk traits in tests. 

my reasoning being, I want to showcase the ease of arithmetic in binary fields and their efficient embedding and small by large multiplication properties. this obviously bloats the code a bit

can remove this entirely, and directly use our traits, if we feel that's the better approach.

**Review by 0xJepsen — APPROVED** [review]
Very nice work on this sir

File: src/field/binary/mod.rs (hunk 2 above)
Comment (Autoparallel) [reply src/field/binary/mod.rs]: What if you use a trait impl override/default/specialization? There are some Rust unstable feats that may help here to let us specialize to the binary case.

I thought we had it in the project, but don't seem to anymore. Anyway, here is an RFC for it:
https://rust-lang.github.io/rfcs/1210-impl-specialization.html

There could be others.

File: src/field/binary_towers/extension.rs [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+/// represented as vector of 2^K [`BinaryField`] components in multilinear basis,
+/// i.e. an element $b_v = \prod_{i=0}^{k-1}(v_i.X_i+(1-v_i))$ where $v_i$ is in GF(2).
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BinaryFieldExtension<const K: usize>
+where [(); 1 << K]: {
+  /// coefficient of field element
+  pub coefficients: [BinaryField; 1 << K],
+}
+
+impl<const K: usize> BinaryFieldExtension<K>
```
Comment (Autoparallel) [inline src/field/binary_towers/extension.rs:26]: Should I be thinking of `BinaryFieldExtension<K>` as the binary tower with height `K`?

**Review by Autoparallel — APPROVED** [review]
This looks good to me! I just had a few clarifying thoughts.

One on a previous comment, one in this review.

File: src/field/binary_towers/extension.rs (hunk 5 above)
Comment (Autoparallel) [reply src/field/binary_towers/extension.rs:26]: Ah nice. Probably a nit, but what do you think of naming this `BinaryTower` so we can think of `GaloisField<K, 2>` as degree `K` extensions over `PrimeField<2>`?

Let it be known that it doesn't bother me either way, you should choose what you think is proper :)

File: src/field/binary/mod.rs (hunk 2 above)
Comment (lonerapier) [reply src/field/binary/mod.rs]: @Autoparallel i need some more time to check and iterate on this. so, i'll open a new PR for this. don't want to block AES implementations due to this change, so merging it.

**Pull request merged**